services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,342 @@
+          // todo: refactor
+          await this.iam.detachRolePolicy({
+            PolicyArn: freshPolicyMetadata.Arn,
+            RoleName: role.RoleName,
+          });
+
+          return undefined;
+        }
+
+        return role;
+      },
+    });
+
+    // INSTANCE PROFILE
+    this.instanceProfile = await readModifySet({
+      read: async () => await this.iam.getInstanceProfile({InstanceProfileName: roleName}),
+      set: async () => await this.iam.createInstanceProfile({InstanceProfileName: roleName}),
+    });
+
+    //todo: do something to enable logging to stackdriver
+  }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0; // why zero?
+      });
+    }
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      ...workerPool.config,
+      running: workerPool.providerData[this.providerId].running,
+    });
+
+    let spawned;
+
+    try {
+      spawned = await this.ec2.runInstances({
+        IamInstanceProfile: {
+          Arn: this.instanceProfile.Arn,
+          Name: this.instanceProfile.InstanceProfileName,
+        },
+        MaxCount: toSpawn,
+        MinCount: toSpawn,
+        ImageId: workerPool.config.imageId,
+        InstanceType: workerPool.config.instanceType,
+        Placement: {
+          AvailabilityZone: this.region,
+        },
+        TagSpecifications: {
+          ResourceType: 'instance',
+          Tags: [{
+            Key: 'Provider',
+            Value: `wm-${this.providerId}`,
+          }, {
+            Key: 'Owner',
+            Value: workerPool.owner,
+          }, {
+            Key: 'Project',
+            Value: this.project,
+          }],
+        },
+      });
+    } catch (e) {
+      await workerPool.reportError({
+        kind: 'creation-error',
+        title: 'Instance Creation Error',
+        description: e.message, // TODO: Make sure we clear exposing this with security folks
+        notify: this.notify,
+        WorkerPoolError: this.WorkerPoolError,
+      });
+    }
+
+    Promise.all(spawned.Instances.map(i => {
+      return this.Worker.create({
+        workerPoolId,
+        providerId: this.providerId,
+        workerGroup: this.providerId,
+        workerId: i.InstanceId,
+        created: new Date(),
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        providerData: { // what do we want to remember, anyways? 🤔
+          project: this.project,
+        },
+      });
+    }));
+  }
+
+  async deprovision({workerPool}) {
+    // should this be implemented in Provider? Looks like it's going to be the same for all
```

Comment:
This is probably wrong for the google provider, too -- I think it should do nothing, so the worker-manager just waits for the instances to terminate.  This changed in https://bugzilla.mozilla.org/show_bug.cgi?id=1555213 and I should have updated the google provider at that time.  I'll make a PR.

Reply:
just to make sure I understand: so we shouldn't delete `providerData` from the db?

Reply:
correct